### ISSUE TITLE
Installing packages that depend on tensorflow break GPU compatibility

### ISSUE THREAD

**Issue by ericjang** [ROCmSoftwarePlatform/tensorflow-upstream#129]
I made a simple extension to the rocm TF docker container:

```
FROM rocm/tensorflow:rocm1.8.2-tf1.8-python2
RUN pip install --user --upgrade tensorflow-probability
RUN pip install matplotlib scipy pandas sklearn
```

installing tensorflow-probability causes the rocm tensorflow installation to be overwritten, which results in failure to schedule on the GPU.

**Comment by whchung** [comment]
@parallelo mind help check the dependency asked by `tensorflow-probability`?

**Comment by parallelo** [comment]
Sure, will take a look now.

**Comment by parallelo** [comment]
While this initially appears to be a pip packaging issue, the fundamental issue looks to be a lack of ROCm target support in the `tensorflow/probability` repository.  

Based on the [README](https://github.com/tensorflow/probability#stable-builds) in `tensorflow/probability`, note that the pre-built `tensorflow-probability` pip package you were trying to use is built only for CPU targets.  There is a separate package called `tensorflow-probability-gpu` that is built for Nvidia GPU targets.

The likely solution would be to add `rocm` flags/targets in addition to the existing `gpu` flags/targets within the `tensorflow/probability` repository.  For future investigation, here's a useful [search result](https://github.com/tensorflow/probability/search?q=gpu&unscoped_q=gpu) of all the mentions of `gpu` in that repo.  

We'll add this to our to-do list, and we also would welcome community contributions to that repo.

**Comment by parallelo** [comment]
Looked a bit more closely at `tensorflow/probability`.  Perhaps the word "target" that I used was too strong.  I'll take a look at this later today, with a goal of adding a ROCm-enabled `tensorflow-rocm` dependency option within that repo.

Useful quote from `tensorflow/probability`:  
> Currently, TensorFlow Probability does not contain any GPU-specific code. The primary difference between these packages is that tensorflow-probability-gpu depends on a GPU-enabled version of TensorFlow.

**Comment by ericjang** [comment]
Thanks for quickly addressing this. I also dug a bit into this issue.

The real pain point is that a number of TF contrib packages (e.g. https://www.tensorflow.org/api_docs/python/tf/contrib/distributions/bijectors) are being split up into separate modules. This creates some version-compatibility issues for ROCM TF port because these "external TF packages" usually depend on latest stable TF release (e.g. 1.10), and it's no longer sufficient to just expect TF to come batteries-included.

The tensorflow_probability package can be successfully installed via 

```pip3 install --no-deps tensorflow_probability``` and this does not override the existing TF installation, but unfortunately the latest version is incompatible with ROCM's TF 1.8.

I just wanted to bring to attention that as increasing TF modules are "split out" of TF, (e.g. T2T, etc.) there might be more confusion on external that explicitly depend on tensorflow-gpu.

**Comment by parallelo** [comment]
Good findings.  Much appreciated.  

Note that if you build `tensorflow-upstream` from source, you'll currently get a TF 1.1.0 whl package.  Today, I'll test one of those pip packages along with `pip3 install --no-deps tensorflow_probability`.

**Comment by ericjang** [comment]
Thanks! Is there a Dockerfile available to build rocm-tf from source?

**Comment by parallelo** [comment]
Good question, I'm not sure if we've released one of our Dockerfiles yet.  Let me check on that, and look into where it would make sense to post one.  

In the meantime, you should be able to build `tensorflow-upstream` from source using the `develop-upstream` branch inside your docker container.  Build steps are posted [here](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/rocm_docs/tensorflow-build-from-source.md).  

I just tried one of the `tensorflow/probability` examples using one of our ROCm TF 1.10 whl files, and it seemed to run without any errors.  (although note that I don't have any previous experience with `tensorflow/probability`)

**Comment by parallelo** [comment]
@ericjang - Keep an eye on my recent PR https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/145.  Please feel free to give it a try and to let us know how it works.

**Comment by ericjang** [comment]
Thanks for the Docker file! I was able to hack a Dockerfile together by building the .whl from source, basing a Docker image off the tf1.8 version, and doing 

```
COPY tensorflow.whl tensorflow.whl
RUN pip3 install tensorflow.whl
```
The exact commands for installation are a bit out of date (folder paths, names of build scripts, etc. ) and I had to checkout the `r1.10-rocm` branch instead of `develop-upstream`. But thankfully that worked and now I can use tensorflow_probability.
